File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-valid-gpu-computation-dimensions.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  GPU_COMPUTATION_HEIGHT_ARGUMENT_INDEX,
  GPU_COMPUTATION_WIDTH_ARGUMENT_INDEX,
  MINIMUM_GPU_COMPUTATION_DIMENSION_PX,
} from "./constants.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { getThreeConstructorName } from "./utils/get-three-constructor-name.js";

export const threeValidGpuComputationDimensions = defineRule({
  id: "three-valid-gpu-computation-dimensions",
  title: "GPU computation has an invalid texture dimension",
  category: "Correctness",
  severity: "error",
  recommendation: "Use positive integer dimensions for GPU computation textures",
  create: (context: RuleContext) => ({
    NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
      if (getThreeConstructorName(node, context.scopes) !== "GPUComputationRenderer") return;
      const dimensions = [
        ["width", node.arguments[GPU_COMPUTATION_WIDTH_ARGUMENT_INDEX]],
        ["height", node.arguments[GPU_COMPUTATION_HEIGHT_ARGUMENT_INDEX]],
      ] as const;
      for (const [dimensionName, argument] of dimensions) {
        if (!argument || isNodeOfType(argument, "SpreadElement")) continue;
        const dimension = getStaticNumber(argument, context.scopes);
        if (
          dimension === null ||
          (Number.isInteger(dimension) && dimension >= MINIMUM_GPU_COMPUTATION_DIMENSION_PX)
        ) {
          continue;
        }
        context.report({
          node: argument,
          message: `GPUComputationRenderer ${dimensionName} must be a positive integer, but this value is ${String(dimension)}`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/three-valid-gpu-computation-dimensions.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, AstNode};

const GPU_COMPUTATION_WIDTH_ARGUMENT_INDEX: usize = 0;
const GPU_COMPUTATION_HEIGHT_ARGUMENT_INDEX: usize = 1;
const MINIMUM_GPU_COMPUTATION_DIMENSION_PX: f64 = 1.0;

#[derive(Debug, Default, Clone)]
pub struct ThreeValidGpuComputationDimensions;

declare_oxc_lint!(
    /// Require positive integer GPU computation texture dimensions.
    ThreeValidGpuComputationDimensions,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Validate GPU computation texture dimensions.",
);

impl Rule for ThreeValidGpuComputationDimensions {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::NewExpression(new_expression) = node.kind() else {
            return;
        };
        if !three_module_api_path_matches(&new_expression.callee, &["GPUComputationRenderer"], ctx)
        {
            return;
        }
        report_invalid_gpu_computation_dimension(
            "width",
            new_expression
                .arguments
                .get(GPU_COMPUTATION_WIDTH_ARGUMENT_INDEX),
            ctx,
        );
        report_invalid_gpu_computation_dimension(
            "height",
            new_expression
                .arguments
                .get(GPU_COMPUTATION_HEIGHT_ARGUMENT_INDEX),
            ctx,
        );
    }
}

fn report_invalid_gpu_computation_dimension<'a>(
    dimension_name: &str,
    argument: Option<&'a oxc_ast::ast::Argument<'a>>,
    ctx: &LintContext<'a>,
) {
    let Some((expression, dimension)) = resolve_static_number_argument(argument, ctx) else {
        return;
    };
    if dimension.fract() == 0.0 && dimension >= MINIMUM_GPU_COMPUTATION_DIMENSION_PX {
        return;
    }
    ctx.diagnostic(
        OxcDiagnostic::warn(format!(
            "GPUComputationRenderer {dimension_name} must be a positive integer, but this value is {dimension}"
        ))
        .with_label(expression.span()),
    );
}
